---
title: "CLAmp Sequencing"
runtime: shiny
output: 
  flexdashboard::flex_dashboard:
    theme: cosmo
    orientation: rows
---

```{r setup, include=FALSE}
library(flexdashboard)
library(knitr)
library(pheatmap)
library(data.table)
library(tidyverse)
library(viridis)
library(RColorBrewer)
library(shiny)
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_knit$set(root.dir= normalizePath('.'))
knitr::opts_chunk$set(error = FALSE)
```

Page 1
===================================== 

Row
-----------------------------------------------------------------------

```{r pressure, echo=FALSE,warning=FALSE,echo = FALSE}
# counts = fread("counts.csv")
# counts = fread("test.csv")
counts = fread("CLAmp8_counts_simpCols.csv")

matrix = fread("clamp_blast_matrix (1).csv") %>% as.data.frame() %>% column_to_rownames("qseqid")
rc = fread("clamp_output_rc.csv")
key= fread("key_clamp.csv") %>% column_to_rownames("Name")
key2= fread("sequences (1).csv") %>% select(Accession,Location,Species_mini) %>% right_join(colnames(matrix) %>% as.data.frame() %>% rename(Accession = ".")) %>% filter(Location!="#N/A") %>% column_to_rownames("Accession")


# mat_colors_col = list(Location = cbind(unique(key2$Location),viridis(length((unique(key2$Location))))))

mat_colors_col2 <- list(Location = c(brewer.pal(9, "Paired")))
    names(mat_colors_col2$Location) <- unique(key2$Location)

xx = pheatmap(t(matrix),annotation_col = key,annotation_row = key2,show_rownames = F,show_colnames = F,
         cutree_cols = 4,annotation_names_row = T,main = "Clamp CV probes vs CV genomes",annotation_colors = mat_colors_col2,
         annotation_legend = T,fontsize = 7)

```

To better characterize the cross-reactivity amongst our clamp-seq CV probes, I performed a blastn search comparing the reverse complement of the 58 designed clamp probes against 565 coronavirus genomes and found 4828 probe - genome matches. Converting the resulting data into a matrix of clamp probes vs genomes, with ppos as the values within the cells, and clustering yields the above heatmap. I pulled some meta info on both the probes and the genomes, distinguishing whether probes were from the inital "general CV" pool or the substrain specific pull. I also pulled a meta column "Geo Location" that was associated with each CV genome.

As a note I did some manual pooling of countries to get a more managable list of Locations and pulled the species associated with each sequence.


Page 2
===================================== 

Row
-----------------------------------------------------------------------

### Chart 1

```{r }
tmp = counts %>% filter(`fasta[, 1]` %in% rownames(matrix)) %>%  column_to_rownames("fasta[, 1]")
tmp = tmp[1:18,]
tmp_sub = tmp %>% select(tmp %>% colSums %>% as.data.frame() %>% rownames_to_column()%>%  filter(. >=10) %>% select(rowname) %>% unlist() %>% as.character())
tmp_sub = tmp_sub %>% rownames_to_column() %>% mutate(rowsum = rowSums(.[2:23])) %>% filter(rowsum>=10) %>% select(-rowsum)

renderTable({

  x = input$id2
  
table = tmp_sub %>% select(c(1,x)) %>% filter(.[,x]>=10)%>% t() %>% as.data.frame()
colnames(table)  = table[1,] %>% unlist()

table = data.frame(table[-1,])
table
   })

```

Row
-----------------------------------------------------------------------

### Chart 2

```{r }
tmp = counts %>% filter(`fasta[, 1]` %in% rownames(matrix)) %>%  column_to_rownames("fasta[, 1]")
tmp = tmp[1:18,]
tmp_sub = tmp %>% select(tmp %>% colSums %>% as.data.frame() %>% rownames_to_column()%>%  filter(. >=10) %>% select(rowname) %>% unlist() %>% as.character())
tmp_sub = tmp_sub %>% rownames_to_column() %>% mutate(rowsum = rowSums(.[2:23])) %>% filter(rowsum>=10) %>% select(-rowsum)

renderTable({

  x = input$id2
  
table = tmp_sub %>% select(c(1,x)) %>% filter(.[,x]>=10)
colnames(table)  = c(x,"read counts")
table
f = matrix %>% rownames_to_column()%>% filter(rowname %in% unlist(table[1])) %>% t() %>% as.data.frame(stringsAsFactors = FALSE) %>%  rownames_to_column()
colnames(f) = unlist(f[1,]) %>% as.character()
f = f[-1,]

f = f %>% left_join(y = key2 %>% rownames_to_column() %>% select(rowname,Species_mini)) %>% filter(!is.na(Species_mini)) %>%select(-rowname)  %>% group_by(Species_mini) %>% arrange_all() %>% filter(row_number() == n())
  })

```

Inputs {.sidebar}
-------------------------------------

### Chart 2

```{r }
selectInput(inputId = "id2", label = "Sample_ID", choices = colnames(tmp))

```



Page 3
===================================== 

Row{data-height=100}
-------------------------------------

### Haplotype

```{r }
hap_key =  fread("Haplotype_key.csv")
haplotype = fread("haplotypes.tsv",sep = "\t",header = F)
tmp = counts %>% filter(`fasta[, 1]` %in% rownames(matrix)) %>%  column_to_rownames("fasta[, 1]")
tmp = tmp[19:58,]

bt <- function(a, b, p = 0.5) {binom.test(a, b, 0.5, alternative=
                            c("greater"), conf.level = 0.95)$p.value}


renderTable({

  x = input$id3

tmp_t = tmp %>% rownames_to_column() %>% select(c(rowname,x)) %>% rename(Clamp_name = "rowname")
colnames(tmp_t)[2] = "Sample"
tmp_t2 = tmp_t %>% right_join(hap_key,) %>% group_by(Pair_id) %>% summarise(total = sum(Sample)) %>% 
  mutate(total = ifelse(total == 0,1,total))
tmp_t3 =  tmp_t %>% right_join(hap_key,) %>% right_join(tmp_t2) %>% mutate(failures = total-Sample)
tmp_t3$p_value = mapply(bt, tmp_t3$Sample, tmp_t3$total)

haplotype_str = tmp_t3 %>% group_by(Pair_id) %>% top_n(-1, p_value) %>% top_n(-1, Clamp_name) %>% mutate(haplotype = ifelse(p_value<.001,Nucleotide,"*"))%>% ungroup %>% select(haplotype) %>% unlist() %>% paste0(collapse = "")

haplotype_str
 })

```

Row {.tabset .tabset-fade} 
-------------------------------------

### Substrain Haplotype best match

```{r test2}

renderTable({

  x = input$id3


tmp_t = tmp %>% rownames_to_column() %>% select(c(rowname,x)) %>% rename(Clamp_name = "rowname")
colnames(tmp_t)[2] = "Sample"
tmp_t2 = tmp_t %>% right_join(hap_key,) %>% group_by(Pair_id) %>% summarise(total = sum(Sample)) %>% 
  mutate(total = ifelse(total == 0,1,total))
tmp_t3 =  tmp_t %>% right_join(hap_key,) %>% right_join(tmp_t2) %>% mutate(failures = total-Sample)
tmp_t3$p_value = mapply(bt, tmp_t3$Sample, tmp_t3$total)

haplotype_str = tmp_t3 %>% group_by(Pair_id) %>% top_n(-1, p_value) %>% top_n(-1, Clamp_name) %>% mutate(haplotype = ifelse(p_value<.001,Nucleotide,"*"))%>% ungroup %>% select(haplotype) %>% unlist() %>% as.data.frame() %>% t()


fixed_split <- function(text, n) {
  strsplit(text, paste0("(?<=.{",n,"})"), perl=TRUE)
}
df <- data.frame(matrix(unlist( fixed_split(haplotype$V2, 1) ), nrow=length(haplotype$V2), byrow=T))
colnames(df) = colnames(haplotype_str)

answer = df == do.call("rbind", replicate(dim(df)[1], haplotype_str, simplify = FALSE))
xxx_2 = apply(haplotype_str == "*", 1, table) %>% t()
xxx = apply(answer, 1, table) %>% t()

haplotype_2=  cbind(haplotype,xxx)
colnames(haplotype_2) = c("Subspecies","Haplotype","Mismatch","Match")
output= haplotype_2 %>% filter(Match + Mismatch <=20) %>% arrange(-Match)  %>% filter(Mismatch == min(Mismatch))
output$Mismatch = output$Mismatch - xxx_2[,2]
output$Match = output$Match
output$`Wild cards` = xxx_2[1,2]
output
 })
```

### Haplotype binomial test calls

```{r test5}

renderTable({

  x = input$id3


tmp_t = tmp %>% rownames_to_column() %>% select(c(rowname,x)) %>% rename(Clamp_name = "rowname")
colnames(tmp_t)[2] = "Sample"
tmp_t2 = tmp_t %>% right_join(hap_key,) %>% group_by(Pair_id) %>% summarise(total = sum(Sample)) %>% 
  mutate(total = ifelse(total == 0,1,total))
tmp_t3 =  tmp_t %>% right_join(hap_key,) %>% right_join(tmp_t2) %>% mutate(failures = total-Sample)
tmp_t3$p_value = mapply(bt, tmp_t3$Sample, tmp_t3$total)

haplotype_str = tmp_t3 %>% group_by(Pair_id) %>% top_n(-1, p_value) %>% top_n(-1, Clamp_name) %>% mutate(haplotype = ifelse(p_value<.001,Nucleotide,"*"))%>% ungroup%>% select(Pair_id,haplotype,Sample,total,p_value,Status) %>% 
  mutate(wt_counts = ifelse(Status == "wildtype",Sample,total-Sample)) %>% 
    mutate(mt_counts = ifelse(Status == "mutant",Sample,total-Sample)) %>% select(-Sample,-total,-Status) %>% 
  select(Pair_id,haplotype,wt_counts,mt_counts,p_value)

haplotype_str$p_value  = formatC(haplotype_str$p_value,format = "G")
haplotype_str
 })
```

Inputs {.sidebar}
-------------------------------------

### Input Column

```{r }
selectInput(inputId = "id3", label = "Sample_ID", choices = colnames(tmp))

```


